### PR TITLE
Allow user to define a password directly via its hash

### PR DESCRIPTION
**Pull request by andrejtonev**
The user can now define their password via a hash.

### CODE REVIEW

**Comment by andrejtonev** [comment]
### Tracking
- [ ] **[Link to Epic/Issue]**


### Standard development
- [ ] Update unit/E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch this branch


### CI Testing Labels
- [ ] Select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_


### Documentation checklist
- [ ] Add the documentation label
- [ ] Add the bug / feature label
- [ ] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **[ Release note text ]**
- [ ] **[ Documentation PR link memgraph/documentation#XXXX ]**
    - [ ] Is back linked to this development PR
- [ ] **[ Tag someone from docs team ]**

**Comment by antejavor** [comment]
Related issue: https://github.com/memgraph/memgraph/issues/2145

File: src/auth/crypto.cpp [hunk 1]
```diff
@@ -258,8 +258,34 @@ PasswordHashAlgorithm &InternalCurrentHashAlgorithm() {
   std::call_once(flag, [] { current = InternalParseHashAlgorithm(FLAGS_password_encryption_algorithm); });
   return current;
 }
+
+std::optional<std::string_view> UsesAlgo(std::string_view str, PasswordHashAlgorithm algo) {
+  // header = algo name + :
+  const auto header = std::string{AsString(algo)} + ":";
```
Comment (Ignition) [inline src/auth/crypto.cpp:264]: The `AsString` responses need to be made absolutely stable, and publicly documented

File: src/auth/crypto.cpp [hunk 2]
```diff
@@ -271,6 +297,16 @@ auto AsString(PasswordHashAlgorithm hash_algo) -> std::string_view {
...
 }
 
+auto HashSize(PasswordHashAlgorithm hash_algo) -> size_t {
+  switch (hash_algo) {
+    case PasswordHashAlgorithm::BCRYPT:
+      return 60;  // NOTE: BCRYPT_HASHSIZE is 64, but the result is actually 60B
+    case PasswordHashAlgorithm::SHA256:
+    case PasswordHashAlgorithm::SHA256_MULTIPLE:
+      return SHA::SHA_LENGTH;
+  }
```
Comment (Ignition) [inline src/auth/crypto.cpp:307]: Return a struct for unsalted vs salted sizes. Or make a something obvious these are unsalted lengths.

File: tests/unit/auth.cpp [hunk 3]
```diff
@@ -352,6 +381,23 @@ TEST_F(AuthWithStorage, UserPasswordCreation) {
     ASSERT_FALSE(auth->Authenticate("test", "456"));
     ASSERT_TRUE(auth->RemoveUser(user->username()));
   }
+
+  {
+    auto user =
+        auth->AddUser("test", "sha256:d74ff0ee8da3b9806b18c877dbf29bbde50b5bd8e4dad7a3a725000feb82e8f1" /* pass */);
```
Comment (Ignition) [inline tests/unit/auth.cpp:387]: Does this mean `CREATE USER 'test' IDENTIFIED BY 'sha256:d74ff0ee8da3b9806b18c877dbf29bbde50b5bd8e4dad7a3a725000feb82e8f1';` would also work?

**Comment by andrejtonev** [comment]
@tonijurjevic96 
I need to resolve PR comments, but we also need to check if your password hiding work here as well.
It should, I just need a reminder to check.

File: tests/unit/auth.cpp (hunk 3 above)
Comment (andrejtonev) [reply tests/unit/auth.cpp:387]: Yes! 
I'll add some e2e tests as well

**Review by Ignition — APPROVED** [review]

**Comment by kgolubic** [comment]
For release note I'll use:

- Users can now define their passwords directly using a hash, allowing for more
secure authentication methods by bypassing the need to store or transmit
plain-text passwords. [#2148](https://github.com/memgraph/memgraph/pull/2148)

**Pull request merged**